diff --git a/Makefile b/Makefile
index 6c62948..b3f284e 100644
--- a/Makefile
+++ b/Makefile
@@ -1,22 +1,33 @@
 AR = /usr/bin/ar
 CC = /usr/bin/gcc
-CFLAGS = -Wall -O3
+CFLAGS = -Wall -O3 -fPIC
+PREFIX := /usr/local
 
 all: hss_lib.a \
      hss_lib_thread.a \
      hss_verify.a \
      demo \
-     test_hss
+     test_hss \
+     libhss.so
 
 hss_lib.a: hss.o hss_alloc.o hss_aux.o hss_common.o \
      hss_compute.o hss_generate.o hss_keygen.o hss_param.o hss_reserve.o \
-     hss_sign.o hss_sign_inc.o hss_thread_single.o \
+     hss_sign.o hss_sign_inc.o hss_thread_pthread.o \
      hss_verify.o hss_verify_inc.o hss_derive.o \
      hss_derive.o hss_zeroize.o lm_common.o \
      lm_ots_common.o lm_ots_sign.o lm_ots_verify.o lm_verify.o endian.o \
      hash.o sha256.o
 	$(AR) rcs $@ $^
 
+libhss.so: hss.o hss_alloc.o hss_aux.o hss_common.o \
+     hss_compute.o hss_generate.o hss_keygen.o hss_param.o hss_reserve.o \
+     hss_sign.o hss_sign_inc.o hss_thread_pthread.o \
+     hss_verify.o hss_verify_inc.o \
+     hss_derive.o hss_zeroize.o lm_common.o \
+     lm_ots_common.o lm_ots_sign.o lm_ots_verify.o lm_verify.o endian.o \
+     hash.o sha256.o
+	$(CC) -shared -o $@ $^
+
 hss_lib_thread.a: hss.o hss_alloc.o hss_aux.o hss_common.o \
      hss_compute.o hss_generate.o hss_keygen.o hss_param.o hss_reserve.o \
      hss_sign.o hss_sign_inc.o hss_thread_pthread.o \
@@ -116,6 +127,30 @@ sha256.o: sha256.c sha256.h endian.h
 	$(CC) $(CFLAGS) -c sha256.c -o $@
 
 clean:
-	-rm *.o *.a demo test_hss
-
-
+	-rm *.so *.o *.a demo test_hss
+
+install: libhss.so
+	install -d $(DESTDIR)$(PREFIX)/include/hss/
+	install -m 755 libhss.so $(DESTDIR)$(PREFIX)/lib/
+	install -m 644 common_defs.h $(DESTDIR)$(PREFIX)/include/hss/
+	install -m 644 config.h $(DESTDIR)$(PREFIX)/include/hss/
+	install -m 644 endian.h $(DESTDIR)$(PREFIX)/include/hss/
+	install -m 644 hash.h $(DESTDIR)$(PREFIX)/include/hss/
+	install -m 644 hss_aux.h $(DESTDIR)$(PREFIX)/include/hss/
+	install -m 644 hss_common.h $(DESTDIR)$(PREFIX)/include/hss/
+	install -m 644 hss_derive.h $(DESTDIR)$(PREFIX)/include/hss/
+	install -m 644 hss.h $(DESTDIR)$(PREFIX)/include/hss/
+	install -m 644 hss_internal.h $(DESTDIR)$(PREFIX)/include/hss/
+	install -m 644 hss_reserve.h $(DESTDIR)$(PREFIX)/include/hss/
+	install -m 644 hss_sign_inc.h $(DESTDIR)$(PREFIX)/include/hss/
+	install -m 644 hss_thread.h $(DESTDIR)$(PREFIX)/include/hss/
+	install -m 644 hss_verify.h $(DESTDIR)$(PREFIX)/include/hss/
+	install -m 644 hss_verify_inc.h $(DESTDIR)$(PREFIX)/include/hss/
+	install -m 644 hss_zeroize.h $(DESTDIR)$(PREFIX)/include/hss/
+	install -m 644 lm_common.h $(DESTDIR)$(PREFIX)/include/hss/
+	install -m 644 lm_ots_common.h $(DESTDIR)$(PREFIX)/include/hss/
+	install -m 644 lm_ots.h $(DESTDIR)$(PREFIX)/include/hss/
+	install -m 644 lm_ots_verify.h $(DESTDIR)$(PREFIX)/include/hss/
+	install -m 644 lm_verify.h $(DESTDIR)$(PREFIX)/include/hss/
+	install -m 644 sha256.h $(DESTDIR)$(PREFIX)/include/hss/
+	install -m 644 test_hss.h $(DESTDIR)$(PREFIX)/include/hss/
diff --git a/sha256.h b/sha256.h
index e4c8a9c..3282c90 100644
--- a/sha256.h
+++ b/sha256.h
@@ -5,7 +5,7 @@
 #include EXT_SHA256_H
 #else
 
-#define USE_OPENSSL 1   /* We use the OpenSSL implementation for SHA-256 */
+#define USE_OPENSSL 0   /* We use the OpenSSL implementation for SHA-256 */
                         /* (which is quite a bit faster than our portable */
                         /* C version) */
 
